wasm/wasm.go

```diff
@@ -233,3 +239,67 @@ func EncodeF64(input float64) uint64 {
 func DecodeF64(input uint64) float64 {
 	return math.Float64frombits(input)
 }
+
+// HostModuleConfig are WebAssembly 1.0 (20191205) exports from the host bound to a module name used by InstantiateHostModule.
+type HostModuleConfig struct {
+	// Name is the module name that these exports can be imported with. Ex. wasi.ModuleSnapshotPreview1
+	Name string
+
+	// Functions adds functions written in Go, which a WebAssembly Module can import.
+	//
+	// The key is the name to export and the value is the func. Ex. WASISnapshotPreview1
+	//
+	// Noting a context exception described later, all parameters or result types must match WebAssembly 1.0 (20191205) value
+	// types. This means uint32, uint64, float32 or float64. Up to one result can be returned.
+	//
+	// Ex. This is a valid host function:
+	//
+	//	addInts := func(x uint32, uint32) uint32 {
+	//		return x + y
+	//	}
+	//
+	// Host functions may also have an initial parameter (param[0]) of type context.Context or wasm.ModuleContext.
+	//
+	// Ex. This uses a Go Context:
+	//
+	//	addInts := func(ctx context.Context, x uint32, uint32) uint32 {
+	//		// add a little extra if we put some in the context!
+	//		return x + y + ctx.Value(extraKey).(uint32)
+	//	}
+	//
+	// The most sophisticated context is wasm.ModuleContext, which allows access to the Go context, but also
+	// allows writing to memory. This is important because there are only numeric types in Wasm. The only way to share other
+	// data is via writing memory and sharing offsets.
+	//
+	// Ex. This reads the parameters from!
+	//
+	//	addInts := func(ctx wasm.ModuleContext, offset uint32) uint32 {
+	//		x, _ := ctx.Memory().ReadUint32Le(offset)
+	//		y, _ := ctx.Memory().ReadUint32Le(offset + 4) // 32 bits == 4 bytes!
+	//		return x + y
+	//	}
+	//
+	// See https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#host-functions%E2%91%A2
+	Functions map[string]interface{}
+
+	// TODO
+	Globals map[string]*HostModuleConfigGlobal
```

Comment:
I had hoped to hide how the current global configuration was implemented as it is messy. That's why I didn't implement #279, yet. I really had hoped we could defer this, and I hope you can reconsider conflating this with lock hunting. You can see you are already tempted to conflating internal details with the public config api just to do both at the same time. It is far better to separate the topics. public api can and probably should have nothing to do with the current internal impl. rushing a public api is a bad idea.

The main idea I had but didn't implement yet was that the user shouldn't have any of this mess in their code. Simply they can configure either a global or a mutable global. so that means two maps. `map[string]interface{}`: `Globals` and `MutableGlobals` this makes it very hard to get wrong. the value is only one of the existing supported value types which we already have code to convert from go types. the value of the ex. uint64 value is the initial value. There's no difference between unset and zero we don't support that. This makes the only weird case besides someone passing the wrong type (which they already can do in `Functions` so that's nothing new) is declaring the same export name in both Globals and MutableGlobals. that's easy enough to document so that's the best I could do with no less than 4 days thinking about this field.

Reply:
ps the `Get` `Set` API of globals won't look like ^^ for the same reason the API to access exported functions don't look like how the host defines them. Each public type especially the even more weird "table" should be done in a different PR. Internals can change 4times a day and we shouldn't care, but we must care about the public API, the what, how, why and for who on each and every symbol